File: ruby/hash.rb
# :specimen: hash.general #
	# :general: Create a hash #
		hash = {:apples => 4, 'oranges' => 8}
		hash = Hash[:apples, 4, 'oranges', 8]
		hash = Hash.new
	# end #
# end #

# :specimen: hash.manipulation #
	# :manipulation: Setting hash elements' values #
		hash = {:apples => 4, 'oranges' => 8}
		hash[:apples] += 3 # => 7
		hash['oranges'] += hash[:apples] # => 15
	# end #
	
	# :manipulation: Add an element to a hash #
		hash = {:apples => 4, 'oranges' => 8}
		hash[:bananas] = 6
	# end #
	
	# :manipulation: Checking if a hash contains a certain key #
		hash = {:apples => 4, 'oranges' => 8}
		hash.has_key? :apples # => true
		hash.has_key? 'grapes' # => false
	# end #
	
	# :manipulation: Delete a hash element #
		hash = {:apples => 4, 'oranges' => 8}
		hash.delete :apples # => {'oranges' => 8}
	# end #
	
	# :manipulation: Traverse a hash #
		hash = {:apples => 4, 'oranges' => 8}
		hash.each do |el|
			
		end
	# end #
	
	# :manipulation: Inverting a hash #
		hash = {:apples => 4, 'oranges' => 8}
		hash.invert # => {4 => :apples, 8 => 'oranges'}
	# end #
	
	# :manipulation: Merge two hashes #
		hash_a = {:apples => 4, 'oranges' => 8}
		hash_b = {:bananas => 5}
		merged = hash_a.clone.update hash_b # => {:apples => 4, 'oranges' => 8, :bananas => 5}
	# end #
# end #

File: php/hash.php
<?php

// :specimen: hash.general //
	// :general: Create a hash //
	  // In PHP, arrays are just hashes where the keys happen to be
	  // ordered integers starting at 0. Both are created with the Array()
	  // method. Interestingly, this means that hashes are ordered. Oh, PHP.
		$foo = Array(); // Creates an empty hash/array
		$bar = Array("key" => "value", "pi" => 3.14); // Initializes a hash
	// end //
// end //

// :specimen: hash.manipulation //
	// :manipulation: Setting hash elements' values //
		$foo["x"] = 12;
		$foo["x"] = "bar";
	// end //
	
	// :manipulation: Add an element to a hash //
		$foo[9] = "baz";
		// To add an element to the hash with the next sequential key, use
		$foo[] = "spam";
	// end //
	
	// :manipulation: Checking if a hash contains a certain key //
	  if($foo["key"]) {
	    // $foo contains key "key"
	  }
	// end //
	
	// :manipulation: Delete a hash element //
	  unset($foo["bar"]); // Removes key/value pair with key "bar"
	// end //
	
	// :manipulation: Traverse a hash //
		foreach($foo as $key => $value) {
		  // Do something
		}
	// end //
	
	// :manipulation: Inverting a hash //
	  // This isn't completely sound -- if a value occurs multiple times,
	  // the last key is used as it's value, and all the others are *lost*.
	  // Note: for this to work properly, all the values have to be legal hash
	  // keys -- i.e., strings or integers.
	  $foo = array_flip($foo); // => Hash with values => keys.
	// end //
	
	// :manipulation: Merge two hashes //
	  // Appends the key/value pairs in $a to the end of $foo.
	  $a = Array("key" => "value");
	  $foo = array_merge($foo,$a);
	// end //
// end //

?>
